File: src/vip/monitor.py
"""Monitor VIP profiles for changes."""

from __future__ import annotations

import json
from datetime import datetime
from pathlib import Path

from vip.config import get_profiles_dir, CHANGELOG_FILE
from vip.fetchers import search as search_fetcher
from vip.fetchers import twitter as twitter_fetcher
from vip.profile import list_profiles, load_profile, save_profile
from vip.resolver import parse_twitter_handle, parse_linkedin_url
from vip.synthesizer import synthesize_profile, detect_changes


def _extract_metadata(content: str) -> dict:
    """Extract metadata (links, name) from existing profile Markdown."""
    import re
    metadata = {"twitter_handle": None, "linkedin_url": None, "name": None}

    name_match = re.search(r"^# (.+)$", content, re.MULTILINE)
    if name_match:
        metadata["name"] = name_match.group(1)

    twitter_match = re.search(r"twitter\.com/(\w+)", content, re.IGNORECASE)
    if not twitter_match:
        twitter_match = re.search(r"x\.com/(\w+)", content, re.IGNORECASE)
    if twitter_match:
        metadata["twitter_handle"] = twitter_match.group(1)

    linkedin_match = re.search(r"(https?://[^/]*linkedin\.com/in/[^\s)]+)", content)
    if linkedin_match:
        metadata["linkedin_url"] = linkedin_match.group(1)

    return metadata


def _gather_fresh_data(metadata: dict) -> tuple[str, list[str]]:
    """Re-fetch data for a person based on profile metadata."""
    raw_parts = []
    sources = []

    name = metadata.get("name", "")

    # Twitter data via bird CLI
    handle = metadata.get("twitter_handle")
    if handle:
        twitter_data = twitter_fetcher.fetch_profile(handle)
        if twitter_data and twitter_data.raw_output:
            raw_parts.append(f"=== Twitter (@{handle}) ===\n{twitter_data.raw_output}")
            sources.append(f"https://twitter.com/{handle}")

    # Web search
    if name:
        results = search_fetcher.search_person(name)
        for r in results:
            raw_parts.append(f"{r.title}\n{r.body}")
            sources.append(r.url)

    return "\n\n".join(raw_parts), sources


def append_changelog(entry: dict) -> None:
    """Append an entry to the changelog file."""
    CHANGELOG_FILE.parent.mkdir(parents=True, exist_ok=True)
    with open(CHANGELOG_FILE, "a") as f:
        f.write(json.dumps(entry, ensure_ascii=False) + "\n")


def read_changelog(days: int = 30) -> list[dict]:
    """Read recent changelog entries."""
    if not CHANGELOG_FILE.exists():
        return []

    entries = []
    cutoff = datetime.now().timestamp() - (days * 86400)

    with open(CHANGELOG_FILE) as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            try:
                entry = json.loads(line)
                ts = datetime.fromisoformat(entry.get("timestamp", "2000-01-01")).timestamp()
                if ts >= cutoff:
                    entries.append(entry)
            except (json.JSONDecodeError, ValueError):
                continue

    return entries


def unread_count() -> int:
    """Count changelog entries from the last 7 days."""
    return len(read_changelog(days=7))


def run_monitor(profiles_dir: Path | None = None, verbose: bool = False) -> list[dict]:
    """Run a full monitoring cycle: check all profiles for updates.

    Returns list of change entries.
    """
    if profiles_dir is None:
        profiles_dir = get_profiles_dir()

    changes = []
    profiles = list_profiles(profiles_dir)

    for p in profiles:
        name = p["name"]
        slug = p["slug"]

        if verbose:
            print(f"  Checking {name}...")

        old_content = load_profile(slug, profiles_dir)
        if not old_content:
            continue

        metadata = _extract_metadata(old_content)
        new_data, sources = _gather_fresh_data(metadata)

        if not new_data.strip():
            if verbose:
                print(f"    No new data found, skipping.")
            continue

        # Detect changes
        change_summary = detect_changes(old_content, new_data)

        if change_summary:
            # Re-synthesize profile
            new_profile = synthesize_profile(new_data, sources)
            save_profile(name, new_profile, profiles_dir)

            entry = {
                "timestamp": datetime.now().isoformat(),
                "name": name,
                "slug": slug,
                "summary": change_summary,
            }
            append_changelog(entry)
            changes.append(entry)

            if verbose:
                print(f"    Changes detected: {change_summary}")
        elif verbose:
            print(f"    No significant changes.")

    return changes

File: lib/monitor.js
import fs from 'fs';
import { CHANGELOG_FILE, getProfilesDir } from './config.js';
import { searchPerson } from './fetchers/search.js';
import * as twitter from './fetchers/twitter.js';
import { listProfiles, loadProfile, saveProfile } from './profile.js';
import { synthesizeProfile, detectChanges } from './synthesizer.js';

export function extractMetadata(content) {
  const meta = { twitterHandle: null, linkedinUrl: null, name: null };

  const nameMatch = content.match(/^# (.+)$/m);
  if (nameMatch) meta.name = nameMatch[1];

  let twMatch = content.match(/twitter\.com\/(\w+)/i);
  if (!twMatch) twMatch = content.match(/x\.com\/(\w+)/i);
  if (twMatch) meta.twitterHandle = twMatch[1];

  const liMatch = content.match(/(https?:\/\/[^/]*linkedin\.com\/in\/[^\s)]+)/);
  if (liMatch) meta.linkedinUrl = liMatch[1];

  return meta;
}

function gatherFreshData(meta) {
  const rawParts = [];
  const sources = [];

  if (meta.twitterHandle) {
    const data = twitter.fetchProfile(meta.twitterHandle);
    if (data?.rawOutput) {
      rawParts.push(`=== Twitter (@${meta.twitterHandle}) ===\n${data.rawOutput}`);
      sources.push(`https://twitter.com/${meta.twitterHandle}`);
    }
  }

  if (meta.name) {
    const results = searchPerson(meta.name);
    for (const r of results) {
      rawParts.push(`${r.title}\n${r.body}`);
      sources.push(r.url);
    }
  }

  return [rawParts.join('\n\n'), sources];
}

export function appendChangelog(entry) {
  const dir = CHANGELOG_FILE.substring(0, CHANGELOG_FILE.lastIndexOf('/'));
  fs.mkdirSync(dir, { recursive: true });
  fs.appendFileSync(CHANGELOG_FILE, JSON.stringify(entry) + '\n');
}

export function readChangelog(days = 30) {
  if (!fs.existsSync(CHANGELOG_FILE)) return [];

  const cutoff = Date.now() - days * 86400000;
  const lines = fs.readFileSync(CHANGELOG_FILE, 'utf-8').split('\n').filter(Boolean);

  return lines
    .map(line => { try { return JSON.parse(line); } catch { return null; } })
    .filter(e => e && new Date(e.timestamp).getTime() >= cutoff);
}

export function unreadCount() {
  return readChangelog(7).length;
}

export async function runMonitor(profilesDir, verbose = false) {
  const dir = profilesDir || getProfilesDir();
  const changes = [];
  const profiles = listProfiles(dir);

  for (const p of profiles) {
    if (verbose) console.log(`  Checking ${p.name}...`);

    const oldContent = loadProfile(p.slug, dir);
    if (!oldContent) continue;

    const meta = extractMetadata(oldContent);
    const [newData, sources] = gatherFreshData(meta);

    if (!newData.trim()) {
      if (verbose) console.log('    No new data found, skipping.');
      continue;
    }

    const changeSummary = await detectChanges(oldContent, newData);

    if (changeSummary) {
      const newProfile = await synthesizeProfile(newData, sources);
      saveProfile(p.name, newProfile, dir);

      const entry = {
        timestamp: new Date().toISOString(),
        name: p.name,
        slug: p.slug,
        summary: changeSummary,
      };
      appendChangelog(entry);
      changes.push(entry);

      if (verbose) console.log(`    Changes detected: ${changeSummary}`);
    } else if (verbose) {
      console.log('    No significant changes.');
    }
  }

  return changes;
}
